pkg/planner/core/rule_decorrelate.go

```diff
@@ -470,6 +510,133 @@ func (*DecorrelateSolver) Name() string {
 	return "decorrelate"
 }
 
+// isJoinKeyUniqueKey checks if join key is unique key.
+// Returns true if the join key forms a unique key constraint.
+func isJoinKeyUniqueKey(apply *logicalop.LogicalApply, plan base.LogicalPlan) bool {
+	var hasMultiRowOperator func(base.LogicalPlan) bool
+	hasMultiRowOperator = func(p base.LogicalPlan) bool {
+		// Check if current node is a JOIN (excluding the outer Apply which is already a Join)
+		// todo: in/exists can also been evaluated to one row, but we don't handle it yet.
+		if _, ok := p.(*logicalop.LogicalJoin); ok {
+			return true
+		}
+		// Check if current node is UNION ALL
+		if _, ok := p.(*logicalop.LogicalUnionAll); ok {
+			return true
+		}
+		// Recursively check children
+		for _, child := range p.Children() {
+			if hasMultiRowOperator(child) {
+				return true
+			}
+		}
+		return false
+	}
+	if hasMultiRowOperator(plan) {
+		return false
+	}
+
+	// Extract join keys from Selection conditions and their children recursively
+	// Join conditions may be pushed down to DataSource or nested in child Selection nodes
+	innerJoinKeys := make([]*expression.Column, 0)
+
+	// Recursively extract all conditions from Selection nodes and their children
+	var extractConditions func(base.LogicalPlan)
+	extractConditions = func(p base.LogicalPlan) {
+		if sel, ok := p.(*logicalop.LogicalSelection); ok {
+			// Check conditions directly on Selection
+			for _, cond := range sel.Conditions {
+				if decExpr := apply.DeCorColFromEqExpr(cond); decExpr != nil {
+					if sf, ok := decExpr.(*expression.ScalarFunction); ok && sf.FuncName.L == ast.EQ {
+						args := sf.GetArgs()
+						if len(args) == 2 {
+							if innerCol, ok := args[1].(*expression.Column); ok {
```

Comment:
Can we guarantee the `args[1]` rather than `args[0]` is the column from the inner side?

Reply:
DeCorColFromEqExpr preserves the order of args[1], which is on the inner side.